### PR TITLE
🚨🚨🚨 `[NLLB Tokenizer]` Fix the prefix tokens 🚨🚨🚨

### LINKED ISSUE #19943: NllbTokenizer/NllbTokenizerFast inserts language code incorrectly when tokenizing target text
### System Info

- `transformers` version: 4.23.1
- Platform: Windows-10-10.0.19044-SP0
- Python version: 3.8.10
- Huggingface_hub version: 0.10.1
- PyTorch version (GPU?): 1.10.1+cu111 (True)
- Tensorflow version (GPU?): 2.7.3 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@LysandreJik @sgugger @SaulLu

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

According to the [documentation](https://huggingface.co/docs/transformers/model_doc/nllb#transformers.NllbTokenizer) for `NllbTokenizer`,

> The tokenization method is `<tokens> <eos> <language code>` for source language documents, and `<language code> <tokens> <eos>` for target language documents.

When you tokenize target text, it incorrectly inserts the language code at the end of the sentence instead of the beginning.

```
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("facebook/nllb-200-distilled-600M")
tokenizer.tgt_lang = "eng_Latn"
article = "UN Chief says there is no military solution in Syria"
tokens = tokenizer(text_target=article).tokens()
```
`tokens` has the value:
```
['▁UN', '▁Chief', '▁says', '▁there', '▁is', '▁no', '▁military', '▁solution', '▁in', '▁Syria', '</s>', 'eng_Latn']
```

### Expected behavior

`tokens` should have the value:
```
['eng_Latn', '▁UN', '▁Chief', '▁says', '▁there', '▁is', '▁no', '▁military', '▁solution', '▁in', '▁Syria', '</s>']
```

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
The NLLB tokenizer's suffix and prefix token were wrong w.r.t to the paper. 
This breaking change fixes the tokenizer. 
Could be none breaking if we add these to the configuration file maybe? But it is a required change
Have to update the tests but should be good.

The big problem was the `prefix` and `suffix` tokens. 
The previous version adds `[self.eos_token_id, self.cur_lang_code]`  at the end of the token sequence for both target and source tokenization. This is wrong as the `NLLB` paper mentions (page 48, 6.1.1. Model Architecture) : 
> Note that we prefix the source sequence with the source language, as opposed to the target
language as previously done in several works (Arivazhagan et al., 2019; Johnson et al.,
2017). This is primarily because we prioritize optimizing zero-shot performance of our
model on any pair of 200 languages at a minor cost to supervised performance.

Previous behaviour:
```python 
>>> from transformers import NllbTokenizer
>>> tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M")
>>> tokenizer("How was your day?").input_ids
[13374, 1398, 4260, 4039, 248130, 2, 256047]
>>> # 2: '</s>'
>>> # 256047 : 'eng_Latn'
```
New behaviour 
```python 
>>> from transformers import NllbTokenizer
>>> tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M")
>>> tokenizer("How was your day?").input_ids
[256047, 13374, 1398, 4260, 4039, 248130, 2]
```

Enabling the old behaviour:

```python 
>>> from transformers import NllbTokenizer
>>> tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M", legacy_behaviour = True)
```

This parameter should be part of the `tokenizer_config.json`.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Thank you for the implementation, @ArthurZucker!

Could you please put in the PR description the gist of the breaking change with a code sample, and how to revert to the previous behavior if users would like that?

Thank you

**Review by sgugger — COMMENTED** [review]
As @LysandreJik said, we would need a way to enable the old behavior for users who rely on it.

**Comment by ArthurZucker** [comment]
Indeed thanks for the tip on how to enable that swiftly!

**Review by sgugger — APPROVED** [review]
Perfect, thanks!

**Comment by ArthurZucker** [comment]
One test is failing with NLLB (running  slow ones locally, `test_encode_decode_with_spaces`), fixing this before merging

**Comment by LysandreJik** [comment]
Cool, I like the flag :)

Can the doc be shown more prominently? Maybe to replace the disclaimer mentioning to tag me? A disclaimer mentioning that we changed it to what it is now, with the code snippet?



image

File: docs/source/en/model_doc/nllb.mdx [hunk 1]
```diff
@@ -12,8 +12,43 @@ specific language governing permissions and limitations under the License.
...
+**DISCLAIMER:** The default behaviour for the tokenizer has recently been fixed (and thus changed! 
+
+The previous version adds [self.eos_token_id, self.cur_lang_code] at the end of the token sequence for both target and source tokenization. This is wrong as the NLLB paper mentions (page 48, 6.1.1. Model Architecture) :
+
+Note that we prefix the source sequence with the source language, as opposed to the target
+language as previously done in several works (Arivazhagan et al., 2019; Johnson et al.,
+2017). This is primarily because we prioritize optimizing zero-shot performance of our
+model on any pair of 200 languages at a minor cost to supervised performance.
+
+Previous behaviour:
```
Comment (sgugger) [inline docs/source/en/model_doc/nllb.mdx:24]: Needs proofing :-)

File: docs/source/en/model_doc/nllb.mdx [hunk 2]
```diff
@@ -12,8 +12,43 @@ specific language governing permissions and limitations under the License.
 
 # NLLB
 
-**DISCLAIMER:** If you see something strange, file a [Github Issue](https://github.com/huggingface/transformers/issues/new?assignees=&labels=bug&template=bug-report.yml) and assign
-@LysandreJik
+**DISCLAIMER:** The default behaviour for the tokenizer has recently been fixed (and thus changed!
```
Comment (LysandreJik) [inline docs/source/en/model_doc/nllb.mdx:15]:
```suggestion
**DISCLAIMER:** The default behaviour for the tokenizer has recently been fixed (and thus changed)!
```

File: docs/source/en/model_doc/nllb.mdx [hunk 3]
```diff
@@ -12,8 +12,43 @@ specific language governing permissions and limitations under the License.
 
 # NLLB
 
-**DISCLAIMER:** If you see something strange, file a [Github Issue](https://github.com/huggingface/transformers/issues/new?assignees=&labels=bug&template=bug-report.yml) and assign
-@LysandreJik
+**DISCLAIMER:** The default behaviour for the tokenizer has recently been fixed (and thus changed! 
+
+The previous version adds [self.eos_token_id, self.cur_lang_code] at the end of the token sequence for both target and source tokenization. This is wrong as the NLLB paper mentions (page 48, 6.1.1. Model Architecture) :
```
Comment (LysandreJik) [inline docs/source/en/model_doc/nllb.mdx:17]:
```suggestion
The previous version adds `[self.eos_token_id, self.cur_lang_code]` at the end of the token sequence for both target and source tokenization. This is wrong as the NLLB paper mentions (page 48, 6.1.1. Model Architecture) :
```

File: docs/source/en/model_doc/nllb.mdx [hunk 4]
```diff
@@ -12,8 +12,43 @@ specific language governing permissions and limitations under the License.
...
+```
+New behaviour
+
+```python
+>>> from transformers import NllbTokenizer
+
+>>> tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M")
+>>> tokenizer("How was your day?").input_ids
+[256047, 13374, 1398, 4260, 4039, 248130, 2]
+Enabling the old behaviour:
```
Comment (LysandreJik) [inline docs/source/en/model_doc/nllb.mdx:44]:
``````suggestion
[256047, 13374, 1398, 4260, 4039, 248130, 2]
 ```

Enabling the old behaviour can be done as follows:
```python
``````

File: docs/source/en/model_doc/nllb.mdx [hunk 5]
```diff
@@ -12,8 +12,43 @@ specific language governing permissions and limitations under the License.
...
+
+>>> from transformers import NllbTokenizer
+
+>>> tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M", legacy_behaviour=True)
+```
+
+For more details, feel free to check the linked [PR](https://github.com/huggingface/transformers/pull/22313) and [Issue](https://github.com/huggingface/transformers/issues/19943)
```
Comment (LysandreJik) [inline docs/source/en/model_doc/nllb.mdx:51]:
```suggestion
For more details, feel free to check the linked [PR](https://github.com/huggingface/transformers/pull/22313) and [Issue](https://github.com/huggingface/transformers/issues/19943).
```

**Review by LysandreJik — APPROVED** [review]
Thank you @ArthurZucker!

**Comment by ArthurZucker** [comment]
Thanks both for proof reading! 👍🏻

**Pull request merged**